### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by RiccardoGrigoletto**
Populated documentation with intros, quick starts, programming style and code flow.

### CODE REVIEW

File: docs/ablation/intro.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
+# Quick Start
+
+Ablation studies have become best practice in Machine Learning research as they provide insights into the relative
```
Comment (jimdowling) [inline docs/ablation/intro.md]: best practice in evaluating model architectures, as they ...

File: docs/ablation/intro.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+Ablation studies have become best practice in Machine Learning research as they provide insights into the relative 
+contribution of the different architectural and regularization components to the performance of models. 
+An ablation study consists of several trials, where one trial could be, 
+e.g., removing the last convolutional layer of a CNN model, retraining the model, 
+and observing the resulting performance. 
+However, as machine learning architectures become ever deeper and data sizes keep growing, 
+there is an explosion in the number of different architecture combinations that need to be evaluated to understand 
+their relative performance. 
+
+Our framework provides a declarative way to define ablation experiments for deep learning model architectures
```
Comment (jimdowling) [inline docs/ablation/intro.md]: Our framework  -> Maggy

Comment (jimdowling) [reply docs/ablation/intro.md]: not just deep learning model architectures

File: docs/ablation/intro.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+An ablation study consists of several trials, where one trial could be, 
+e.g., removing the last convolutional layer of a CNN model, retraining the model, 
+and observing the resulting performance. 
+However, as machine learning architectures become ever deeper and data sizes keep growing, 
+there is an explosion in the number of different architecture combinations that need to be evaluated to understand 
+their relative performance. 
+
+Our framework provides a declarative way to define ablation experiments for deep learning model architectures 
+and training datasets, in a way that eliminates the need for maintaining redundant copies of code for an ablation study.
+Furthermore, our framework enables parallel execution of ablation trials without requiring the developers to
```
Comment (jimdowling) [inline docs/ablation/intro.md:14]: our framework -> Maggy

File: docs/ablation/intro.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+In order to use Maggy for Ablation studies, first we need to define a model generator function.
+```py
+from maggy.ablation import AblationStudy
+
+ablation_study = AblationStudy('titanic_train_dataset', training_dataset_version=1,
+                              label_name='survived')
+```
+
+### Add the data features to ablate
```
Comment (jimdowling) [inline docs/ablation/intro.md]: Add the features to ablate
(features are data)

File: docs/ablation/intro.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+ablation_study.features.include('pclass')
+
+# include a list of features
+list_of_features = ['fare', 'sibsp']
+ablation_study.features.include(list_of_features)
+```
+
+### Add the model layers to ablate
+
+By model ablation we mean removing the components of the model and observing the resulting performance.
```
Comment (jimdowling) [inline docs/ablation/intro.md]: components of the model, retraining, and observing...

File: docs/ablation/intro.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+# include some single layers in the ablation study
+
+ablation_study.model.layers.include('my_dense_two', 
+                                    'my_dense_three', 
+                                    'my_dense_four', 
+                                    'my_dense_sigmoid'
+                                    )
+```
+
+For simple, not-so-deep models, it might make sense to add the layers to the model one-by-one.
```
Comment (jimdowling) [inline docs/ablation/intro.md]: simple, not-so-deep models, -> For smaller models

File: docs/ablation/intro.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+ablation_study.model.layers.include('my_dense_two', 
+                                    'my_dense_three', 
+                                    'my_dense_four', 
+                                    'my_dense_sigmoid'
+                                    )
+```
+
+For simple, not-so-deep models, it might make sense to add the layers to the model one-by-one. 
+However, in many well-known neural network architectures, we have tens or hundreds of layers that sometimes come
```
Comment (jimdowling) [inline docs/ablation/intro.md]: well-known neural network architectures, such as Transformers,

File: docs/ablation/intro.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+# add a layer group using a list
+ablation_study.model.layers.include_groups(['my_dense_two', 'my_dense_four'])
+
+# add a layer group using a prefix
+ablation_study.model.layers.include_groups(prefix='my_dense')
+```
+
+### Wrap the training logic
+
+Now the only thing you need to do is to wrap your training code in a Python function.
```
Comment (jimdowling) [inline docs/ablation/intro.md]: "wrap your training code in a Python function"
  I don't actually like this phrase, i know Moritz used it.
  
  I would write  as the header
  "Write your training logic"
  then
  
the only thing you need to do is to write your training code in a Python function.

File: docs/ablation/intro.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+ablation_study.model.layers.include_groups(['my_dense_two', 'my_dense_four'])
+
+# add a layer group using a prefix
+ablation_study.model.layers.include_groups(prefix='my_dense')
+```
+
+### Wrap the training logic
+
+Now the only thing you need to do is to wrap your training code in a Python function. 
+You can name this function whatever you wish, but we will refer to it as the training or wrapper function.
```
Comment (jimdowling) [inline docs/ablation/intro.md]: Let's just call it the "training function" to keep it simple.

File: docs/ablation/intro.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+# add a layer group using a prefix
+ablation_study.model.layers.include_groups(prefix='my_dense')
+```
+
+### Wrap the training logic
+
+Now the only thing you need to do is to wrap your training code in a Python function. 
+You can name this function whatever you wish, but we will refer to it as the training or wrapper function. 
+The model_function and dataset_function used in the code are generated by the ablator per each trial, 
+and you should call them in your code. This is your everyday TensorFlow/Keras code:
```
Comment (jimdowling) [inline docs/ablation/intro.md]: everyday -> run of the mill

File: docs/dist_training/intro.md [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# Introduction
+
+Distributed training is useful for big models that can't fit in a single machine and for very big datasets.
+There are several techniques available. As an example, the Mirrored Strategies replicate the models over the workers and 
+train them using splits of the data. 
+
+With Maggy, you can train a Machine Learning model in a distributed fashion without rewriting your code.
```
Comment (jimdowling) [inline docs/dist_training/intro.md]: without rewriting the code in your training function.

File: docs/dist_training/intro.md [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+Distributed Training with Maggy is available on TensorFlow and PyTorch.
+
+If you want to know more on how to use Maggy for DT, you can watch the presentation in the next section.
```
Comment (jimdowling) [inline docs/dist_training/intro.md]: DT -> Distributed Training

File: docs/dist_training/tensorflow.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+```py
+# Extract the data
+(x_train, y_train),(x_test, y_test) = split_dataset(dataset)
+
+# Do some preprocessing operations
+...
+```
+* Define a training function containing the training logic.
+```py
+def training_function(hparams):
```
Comment (jimdowling) [inline docs/dist_training/tensorflow.md]: rename hparams to config
You call it config below in the lagom method.

File: docs/dist_training/tensorflow.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+                             hparams=model_parameters),
+                             ...
+                             )
+
+experiment.lagom(train_fn=training_function, config=config)
+```
+There are many parameters for the configuration object:
+    * model: A tf.keras.Model superclass or list of them.
+           Note that this has to be the class itself, not an instance.
+    * train_set: The training set for the training function. If you want to load the set
```
Comment (jimdowling) [inline docs/dist_training/tensorflow.md:48]: Define what it is. Is it a Spark DataFrame? Is it a list of Spark DataFrames?

File: docs/hpo/intro.md [hunk 15]
```diff
@@ -0,0 +1,19 @@
+
+# Introduction
+
```
Comment (jimdowling) [inline docs/hpo/intro.md:3]: This is not a great intro.
Maggy is:
* a programming model to simplify distributed machine learning
* a parallel execution engine for distributed machine learning
* an experiment tracking framework, if you run Maggy on Hopsworks

Maggy currently supports two execution engines:
* PySpark for distributed ML across many workers 
* Python for distributed ML, currently on a single worker


### Hello World Python

A hello-world Maggy program in Python is as follows:
def train():
  (...)
  return {"accuracy": 1.0 }
  
from maggy import experiment
experiment.lagom(train)

The above code will run the train function and register an experiment on Hopsworks with the accuracy=1.0 result.

### Hello World PySpark

def pytrain(config):
  iris = datasets.load_iris()
  X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)
  clf = XGBClassifier(max_depth=config.get("max_depth"))
  y_pred = clf.predict(X_test)
  accuracy = metrics.accuracy_score(y_test, y_pred)
  return {"accuracy": accuracy }
  
from maggy import experiment
config = Config ("max_depth" = [3. 4, 5])
experiment.lagom(pytrain, config)


### Maggy for Parallel Hyperparameter Tuning

File: docs/hpo/intro.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+Developers can use an existing Optimizer, such as asynchronous successive halving (ASHA), or provide their own one. 
+The basic approach we followed was to add support for the Driver and Executors to communicate via RPCs. 
+The Optimizer that guides hyperparameter search is located on the Driver, and it assigns trials to Executors. 
+Executors periodically send back to the Driver the current performance of their trial, 
+and the Optimizer can decide to early-stop its ongoing trial, followed by sending the Executor with a new trial. 
+Because of the impedance mismatch between trials and the stage-/task-based execution model of Spark, 
+we are blocking Executors with long-running tasks to run multiple trials per task. 
+In this way, Executors are always kept busy running trials, and global information needed for efficient 
+early-stopping is aggregated in the Optimizer.
+If you want to know more about Maggy for HPO, you can watch the presentation in the video posted below.
```
Comment (jimdowling) [inline docs/hpo/intro.md]: HPO -> Hyperparameter Optimization (HPO)

File: docs/hpo/strategies.md [hunk 17]
```diff
@@ -0,0 +1,163 @@
+# Quick Start
+
+Using maggy for Hyperparameter Optimization (HPO) works as follows:
+
+* Define a training function containing the training logic.
+```py
+def training_function(hparams):
```
Comment (jimdowling) [inline docs/hpo/strategies.md]: Is it config or hparams?

We should be consistent across all examples.
Honestly, i like hparams, but if you have config that is ok, took. Just pick one.

File: docs/hpo/strategies.md [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    #training and testing logic
+    ...
+```
+
+* Define a search space, containing the HPs we want to optimize, their type and range.
```
Comment (jimdowling) [inline docs/hpo/strategies.md]: HPs - hparams

File: docs/hpo/strategies.md [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+```py
+config = OptimizationConfig(num_trials=4, 
+                            optimizer="randomsearch", 
+                            searchspace=sp, 
+                            direction="max", 
+                            es_interval=1, 
+                            es_min=5, 
+                            name="hp_tuning_test")
+
+experiment.lagom(train_fn=training_function, config=config)
```
Comment (jimdowling) [inline docs/hpo/strategies.md:28]: we don't have "train_fn" any more, right?

Comment (jimdowling) [reply docs/hpo/strategies.md:28]: Oh, i see we do. Why do we need to have it as a named parameter? We have to have a training function as the first parameter.

File: docs/hpo/strategies.md [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    * hb_interval: Heartbeat interval with which the server is polling.
+    * model: The class of the model to be used in the training function.
+    * train_set: The train_set to be used in the training function.
+    * test_set: The test_set to be used in the training function.
+
+# Strategies
+
+### Random Search
+
+With Random Search, the HPs are selected randomly within the search space defined. The search space is defined
```
Comment (jimdowling) [inline docs/hpo/strategies.md]: HPs ->hparams
Please fix all instances of this.

File: docs/start/quickstart.md [hunk 21]
```diff
@@ -1 +1,192 @@
 # Quickstart
+
+The programming model consists of wrapping the code containing the model training logic inside a function.
```
Comment (jimdowling) [inline docs/start/quickstart.md]: wrapping the code containing the model training logic inside a function ->
writing a Python function that includes the training logic, that we call the training function.

File: docs/start/quickstart.md [hunk 22]
```diff
@@ -1 +1,192 @@
 # Quickstart
+
+The programming model consists of wrapping the code containing the model training logic inside a function. 
+Inside that wrapper function provide all imports and parts that make up your experiment.
```
Comment (jimdowling) [inline docs/start/quickstart.md]: training function
Please change everywhere.

File: docs/start/quickstart.md [hunk 23]
```diff
@@ -1 +1,192 @@
...
+
+## What can I do with Maggy?
```
Comment (jimdowling) [inline docs/start/quickstart.md:6]: What kind of distributed machine learning support do I get with  Maggy?

File: docs/start/quickstart.md [hunk 24]
```diff
@@ -1 +1,192 @@
...
+
+Maggy can be used for three different ML experiments: Distributed Hyperparameter Optimization (HPO),
```
Comment (jimdowling) [inline docs/start/quickstart.md]: four different

data parallel training, and model-parallel training (with DeepSpeed Zero)

File: docs/start/quickstart.md [hunk 25]
```diff
@@ -1 +1,192 @@
 # Quickstart
+
+The programming model consists of wrapping the code containing the model training logic inside a function. 
+Inside that wrapper function provide all imports and parts that make up your experiment.
+
+## What can I do with Maggy?
+
+Maggy can be used for three different ML experiments: Distributed Hyperparameter Optimization (HPO), 
+Distributed Ablation Study and Distributed Training (DT), their availability depends on which framework
```
Comment (jimdowling) [inline docs/start/quickstart.md]: Data parallel training is currently available on TensorFlow, and Model Parallel training is available with DeepSpeed-Zero on PyTorch.

File: docs/start/quickstart.md [hunk 26]
```diff
@@ -1 +1,192 @@
...
+| **Databricks** | **No**              | N.A.                                            | N.A.                           | Tensorflow               |
+| **Local**      | **Yes**             | N.A.                                            | N.A.                           | Tensorflow, Pytorch      |
+| **Local**      | **No**              | N.A.                                            | N.A.                           | Tensorflow               |
+
+Spark is available automatically when you use Hopsworks and Databricks. Locally, it is easy to 
+[install and set up](https://spark.apache.org/docs/latest/api/python/getting_started/install.html).
+
+## Preparing your project to be used with Maggy
+
+Maggy works by defining a function that contains the training logic of your model.
```
Comment (jimdowling) [inline docs/start/quickstart.md]: Repetition - remove this line.

File: docs/start/quickstart.md [hunk 27]
```diff
@@ -1 +1,192 @@
...
+With Maggy, the code will be like this:
+
+```py
+#datasets preparation
+...
+
+#model definition
+...
+
+def training_function(model, train_set, test_set, params):
```
Comment (jimdowling) [inline docs/start/quickstart.md]: params -> hparams or config

File: docs/start/quickstart.md [hunk 28]
```diff
@@ -1 +1,192 @@
...
+    ...
+    
+    #model testing
+    ...
+
+#configure maggy, pass training_function to maggy and run the experiment
+...
+```
+
+There are three requirements for the wrapper function:
```
Comment (jimdowling) [inline docs/start/quickstart.md]: wrapper...

File: docs/start/quickstart.md [hunk 29]
```diff
@@ -1 +1,192 @@
...
+
+This case is the least efficient for Hyperparameter Optimization and Ablation Studies as it will process the data 
+and build a model for each execution. However, it won't affect Distributed Training.
+
+
+## Simple Example
+
+[Colab Example](https://colab.research.google.com/drive/1ZoLuQbL0Il86hTPQ58ilzEfVF6MWpWRi?usp=sharing)
+
+In this example we are using Maggy for Distributed Training of a simple CNN model for the mnist dataset.
```
Comment (jimdowling) [inline docs/start/quickstart.md]: data parallel training

File: docs/start/quickstart.md [hunk 30]
```diff
@@ -1 +1,192 @@
...
+    def call(self, x):
+        x = self.conv1(x)
+        x = self.flatten(x)
+        x = self.d1(x)
+        return self.d2(x)
+
+model = SimpleCNN #do not initialize the model, Maggy will do it for you
+```
+
+Now we wrap the trianinig logic in a function the we called _training_function_
```
Comment (jimdowling) [inline docs/start/quickstart.md]: trianinig -> training

File: docs/start/quickstart.md [hunk 31]
```diff
@@ -1 +1,192 @@
...
+    loss = model.evaluate(
+        test_set[0],
+        test_set[1],
+        batch_size=32)
+    
+    return loss
+```
+
+Finally, we create a configuration class of the experiment we want to run (the configuration classes are 
+TfDistributedConfig, TorchDistributedConfig, HyperparameterOptConfig and AblationConfig) and we launch
```
Comment (jimdowling) [inline docs/start/quickstart.md:173]: Config, TfDistributedConfig, ...

**Review by jimdowling — CHANGES_REQUESTED** [review]

**Comment by RiccardoGrigoletto** [comment]
@jimdowling Next time maybe it's better if the review of the documentation is done directly from the branch and commit the changes, much easier and effective

File: docs/hpo/intro.md (hunk 15 above)
Comment (RiccardoGrigoletto) [reply docs/hpo/intro.md:3]: Are you describing the intro of Maggy or HPO?

File: docs/hpo/strategies.md (hunk 17 above)
Comment (RiccardoGrigoletto) [reply docs/hpo/strategies.md]: Config is passed to lagom, not to the training function, I think I'll keep it as it is

Comment (RiccardoGrigoletto) [reply docs/hpo/strategies.md]: Actually, I'll change the parameters to  'model, train_set, test_set, hparams'

**Review by jimdowling — APPROVED** [review]
